File: AndroidArchitecture/mvvm-live/src/test/java/com/codearms/maoqiqi/app/statistics/StatisticsViewModelTest.java
package com.codearms.maoqiqi.app.statistics;

import android.arch.core.executor.testing.InstantTaskExecutorRule;

import com.codearms.maoqiqi.app.Event;
import com.codearms.maoqiqi.app.LiveDataTestUtils;
import com.codearms.maoqiqi.app.data.TaskBean;
import com.codearms.maoqiqi.app.data.source.TasksDataSource;
import com.codearms.maoqiqi.app.data.source.TasksRepository;
import com.codearms.maoqiqi.app.utils.MessageMap;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.verify;

/**
 * Unit tests for the implementation of {@link StatisticsViewModel}
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/19 10:18
 */
public class StatisticsViewModelTest {

    @Rule
    public InstantTaskExecutorRule instantExecutorRule = new InstantTaskExecutorRule();

    private final static String TITLE = "TITLE";
    private final static String DESCRIPTION = "DESCRIPTION";

    @Mock
    private TasksRepository tasksRepository;

    @Captor
    private ArgumentCaptor<TasksDataSource.LoadTasksCallBack> loadTasksCallBackArgumentCaptor;

    private StatisticsViewModel statisticsViewModel;

    private List<TaskBean> taskBeanList;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);

        statisticsViewModel = new StatisticsViewModel(tasksRepository);

        taskBeanList = new ArrayList<>();
        taskBeanList.add(new TaskBean(TITLE, DESCRIPTION, false));
        taskBeanList.add(new TaskBean(TITLE, DESCRIPTION, false));
        taskBeanList.add(new TaskBean(TITLE, DESCRIPTION, true));
    }

    @Test
    public void loadStatisticsEmpty() {
        // Given an initialized StatisticsViewModel with no tasks
        taskBeanList.clear();

        statisticsViewModel.start();

        verify(tasksRepository).loadTasks(loadTasksCallBackArgumentCaptor.capture());

        loadTasksCallBackArgumentCaptor.getValue().onTasksLoaded(taskBeanList);

        assertThat(statisticsViewModel.observableActiveTasks.getValue(), is(0));
        assertThat(statisticsViewModel.observableCompletedTasks.getValue(), is(0));
    }

    @Test
    public void loadStatistics() {
        statisticsViewModel.start();

        verify(tasksRepository).loadTasks(loadTasksCallBackArgumentCaptor.capture());

        loadTasksCallBackArgumentCaptor.getValue().onTasksLoaded(taskBeanList);

        assertThat(statisticsViewModel.observableActiveTasks.getValue(), is(2));
        assertThat(statisticsViewModel.observableCompletedTasks.getValue(), is(1));
    }

    @Test
    public void loadStatisticsError() throws InterruptedException {
        statisticsViewModel.start();

        verify(tasksRepository).loadTasks(loadTasksCallBackArgumentCaptor.capture());

        loadTasksCallBackArgumentCaptor.getValue().onDataNotAvailable();

        Event<String> event = LiveDataTestUtils.getValue(statisticsViewModel.getMessage());
        assertEquals(event.getContent(), MessageMap.NO_DATA);
    }
}

File: AndroidArchitecture/mvvm-live-kotlin/src/test/java/com/codearms/maoqiqi/app/statistics/StatisticsViewModelTest.kt
package com.codearms.maoqiqi.app.statistics

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import com.codearms.maoqiqi.app.LiveDataTestUtils
import com.codearms.maoqiqi.app.data.TaskBean
import com.codearms.maoqiqi.app.data.source.TasksDataSource
import com.codearms.maoqiqi.app.data.source.TasksRepository
import com.codearms.maoqiqi.app.utils.MessageMap
import org.hamcrest.core.Is.`is`
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.ArgumentCaptor
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import java.util.*

/**
 * Unit tests for the implementation of [StatisticsViewModel]
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/19 10:18
 */
class StatisticsViewModelTest {

    @get:Rule
    var instantExecutorRule = InstantTaskExecutorRule()

    @Mock
    private lateinit var tasksRepository: TasksRepository

    @Captor
    private lateinit var loadTasksCallBackArgumentCaptor: ArgumentCaptor<TasksDataSource.LoadTasksCallBack>

    private lateinit var statisticsViewModel: StatisticsViewModel

    private lateinit var taskBeanList: MutableList<TaskBean>

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)

        statisticsViewModel = StatisticsViewModel(tasksRepository)

        taskBeanList = ArrayList()
        taskBeanList.add(TaskBean(TITLE, DESCRIPTION, false))
        taskBeanList.add(TaskBean(TITLE, DESCRIPTION, false))
        taskBeanList.add(TaskBean(TITLE, DESCRIPTION, true))
    }

    @Test
    fun loadStatisticsEmpty() {
        // Given an initialized StatisticsViewModel with no tasks
        taskBeanList.clear()

        statisticsViewModel.start()

        verify(tasksRepository).loadTasks(loadTasksCallBackArgumentCaptor.capture())

        loadTasksCallBackArgumentCaptor.value.onTasksLoaded(taskBeanList)

        assertThat(statisticsViewModel.observableActiveTasks.value, `is`(0))
        assertThat(statisticsViewModel.observableCompletedTasks.value, `is`(0))
    }

    @Test
    fun loadStatistics() {
        statisticsViewModel.start()

        verify(tasksRepository).loadTasks(loadTasksCallBackArgumentCaptor.capture())

        loadTasksCallBackArgumentCaptor.value.onTasksLoaded(taskBeanList)

        assertThat(statisticsViewModel.observableActiveTasks.value, `is`(2))
        assertThat(statisticsViewModel.observableCompletedTasks.value, `is`(1))
    }

    @Test
    @Throws(InterruptedException::class)
    fun loadStatisticsError() {
        statisticsViewModel.start()

        verify(tasksRepository).loadTasks(loadTasksCallBackArgumentCaptor.capture())

        loadTasksCallBackArgumentCaptor.value.onDataNotAvailable()

        val event = LiveDataTestUtils.getValue(statisticsViewModel.message)
        assertEquals(event.getContent(), MessageMap.NO_DATA)
    }

    companion object {
        private const val TITLE = "TITLE"
        private const val DESCRIPTION = "DESCRIPTION"
    }
}